src/lib/network_pool/snark_pool.ml

```diff
@@ -252,53 +255,81 @@ struct
+      let handle_new_best_tip_ledger t ledger =
+        let open Mina_base in
+        trace_recurring "handle_new_best_tip_ledger" (fun () ->
+            Throttle.enqueue t.snark_table_lock (fun () ->
+                let%bind prover_account_ids_by_statement =
+                  trace_recurring
+                    "generate account ids of provers in snark table" (fun () ->
+                      let prover_account_id
+                          ({fee= {prover; _}; _} : 'a Priced_proof.t) =
+                        Account_id.create prover Token_id.default
+                      in
+                      let account_ids =
+                        Statement_table.map t.snark_tables.all
+                          ~f:prover_account_id
+                      in
+                      Scheduler.yield () >>| Fn.const account_ids )
                 in
-                if not keep then
-                  Hashtbl.remove t.snark_tables.rebroadcastable key ;
-                keep ) ;
-            return ()
-        | `New_refcount_table
-            { Extensions.Snark_pool_refcount.removed
-            ; refcount_table
-            ; best_tip_table } ->
+                (* if this is still starving the scheduler, we can make `location_of_account_batch` yield while it traverses the masks *)
+                let%map prover_account_locations =
+                  trace_recurring
+                    "lookup prover account locations in best tip ledger"
+                    (fun () ->
+                      let account_locations =
+                        Statement_table.data prover_account_ids_by_statement
+                        |> Base_ledger.location_of_account_batch ledger
+                        |> Account_id.Map.of_alist_exn
+                      in
+                      Scheduler.yield () >>| Fn.const account_locations )
+                in
+                Statement_table.filteri_inplace t.snark_tables.all
+                  ~f:(fun ~key ~data:{fee= {fee; _}; _} ->
+                    let prover_account_exists =
+                      Statement_table.find_exn prover_account_ids_by_statement
+                        key
+                      |> Map.find_exn prover_account_locations
+                      |> Option.is_some
+                    in
+                    let keep =
+                      fee_is_sufficient t ~fee
+                        ~account_exists:prover_account_exists
+                    in
+                    if not keep then
+                      Hashtbl.remove t.snark_tables.rebroadcastable key ;
+                    keep ) ) )
+
+      let handle_new_refcount_table t
+          ({removed; refcount_table; best_tip_table} :
+            Extensions.Snark_pool_refcount.view) =
+        trace_recurring "handle_new_refcount_table" (fun () ->
+            (* I think this is needed here to make the tracing work properly here? TODO: test to see if this is true *)
+            let%map () = Deferred.return () in
```

Comment:
This may not be necessary, so maybe I should just remove it and test without it. But I'm currently confused about whether or not the first step of the computation that occurs outside of the monad in the function passed into `trace_recurring` is actually timed in the async cycle or not. I'll just remove this for now and we will observer how it surfaces in the trace files.